### PR TITLE
[MISC] Config parallel with member thread_count

### PR DESCRIPTION
**Pull request by joergi-w**
This PR adds a member variable `thread_count` to the parallel config (it replaces the old value member). It influences both `seqan3::align_cfg::parallel` and `seqan3::search_cfg::parallel`.

Documentation and tests are included in the second commit.

Closes https://github.com/seqan/product_backlog/issues/68 and contributes to https://github.com/seqan/product_backlog/issues/131.

### CODE REVIEW

File: include/seqan3/core/algorithm/configuration_element_parallel_mode.hpp [hunk 1]
```diff
@@ -25,9 +26,32 @@ namespace seqan3::detail
...
+
+    /*!\brief Sets the number of threads for the parallel configuration element.
+     * \param[in] thread_c The maximum number of threads to be used by the algorithm.
+     */
+    explicit parallel_mode(uint32_t thread_c) noexcept : thread_count{thread_c}
+    {}
+    //!\}
+
+    //!\brief The maximum number of threads the algorithm can use.
+    uint32_t thread_count{1u};
```
Comment (joergi-w) [inline include/seqan3/core/algorithm/configuration_element_parallel_mode.hpp:50]: **Question:** Do we want `std::thread::hardware_concurrency` as a default here?

File: test/unit/alignment/configuration/align_config_parallel_test.cpp [hunk 2]
```diff
@@ -23,17 +36,17 @@ TEST(align_config_parallel, configuration)
     { // from lvalue.
         seqan3::align_cfg::parallel elem{2};
         seqan3::configuration cfg{elem};
-        EXPECT_TRUE((std::is_same_v<std::remove_reference_t<decltype(std::get<seqan3::align_cfg::parallel>(cfg).value)>,
+        EXPECT_TRUE((std::is_same_v<std::remove_reference_t<decltype(std::get<seqan3::align_cfg::parallel>(cfg).thread_count)>,
```
Comment (simonsasse) [inline test/unit/alignment/configuration/align_config_parallel_test.cpp:39]: This line is a bit too long. I don't know where to split it though. 🧐🤔

File: test/unit/alignment/configuration/align_config_parallel_test.cpp [hunk 3]
```diff
@@ -23,17 +36,17 @@ TEST(align_config_parallel, configuration)
...
                                     uint32_t>));
 
-        EXPECT_EQ(std::get<seqan3::align_cfg::parallel>(cfg).value, 2u);
+        EXPECT_EQ(std::get<seqan3::align_cfg::parallel>(cfg).thread_count, 2u);
     }
 
     { // from rvalue.
         seqan3::configuration cfg{seqan3::align_cfg::parallel{2}};
-        EXPECT_TRUE((std::is_same_v<std::remove_reference_t<decltype(std::get<seqan3::align_cfg::parallel>(cfg).value)>,
+        EXPECT_TRUE((std::is_same_v<std::remove_reference_t<decltype(std::get<seqan3::align_cfg::parallel>(cfg).thread_count)>,
```
Comment (simonsasse) [inline test/unit/alignment/configuration/align_config_parallel_test.cpp:47]: Same here. :)

**Comment by joergi-w** [comment]
Rebased on master.

File: include/seqan3/core/algorithm/configuration_element_parallel_mode.hpp [hunk 4]
```diff
@@ -25,9 +26,32 @@ namespace seqan3::detail
...
+    parallel_mode(parallel_mode const &) = default; //!< Defaulted.
+    parallel_mode(parallel_mode &&) noexcept = default; //!< Defaulted.
+    parallel_mode & operator=(parallel_mode const &) = default; //!< Defaulted.
+    parallel_mode & operator=(parallel_mode &&) noexcept = default; //!< Defaulted.
+    ~parallel_mode() = default; //!< Defaulted.
+
+    /*!\brief Sets the number of threads for the parallel configuration element.
+     * \param[in] thread_c The maximum number of threads to be used by the algorithm.
+     */
+    explicit parallel_mode(uint32_t thread_c) noexcept : thread_count{thread_c}
```
Comment (smehringer) [inline include/seqan3/core/algorithm/configuration_element_parallel_mode.hpp:45]:
```suggestion
     * \param[in] thread_count_ The maximum number of threads to be used by the algorithm.
     */
    explicit parallel_mode(uint32_t thread_count_) noexcept : thread_count{thread_count_}

:nail_care:

File: include/seqan3/core/algorithm/configuration_element_parallel_mode.hpp [hunk 5]
```diff
@@ -25,9 +26,32 @@ namespace seqan3::detail
...
-    //!\brief Internal id to check for consistent configuration settings.
+public:
+    /*!\name Constructors, assignment and destructor
+     * \{
+     */
+    parallel_mode() = default; //!< Defaulted.
+    parallel_mode(parallel_mode const &) = default; //!< Defaulted.
+    parallel_mode(parallel_mode &&) noexcept = default; //!< Defaulted.
+    parallel_mode & operator=(parallel_mode const &) = default; //!< Defaulted.
+    parallel_mode & operator=(parallel_mode &&) noexcept = default; //!< Defaulted.
```
Comment (smehringer) [inline include/seqan3/core/algorithm/configuration_element_parallel_mode.hpp:39]:
```suggestion
    parallel_mode(parallel_mode &&) = default; //!< Defaulted.
    parallel_mode & operator=(parallel_mode const &) = default; //!< Defaulted.
    parallel_mode & operator=(parallel_mode &&) = default; //!< Defaulted.
```
or did you define this on purpose? because usually for `= default` constructors this is added automatically.

File: include/seqan3/core/algorithm/configuration_element_parallel_mode.hpp (hunk 1 above)
Comment (smehringer) [reply include/seqan3/core/algorithm/configuration_element_parallel_mode.hpp:50]: hmm good question. I would vote for yes. I'll ask @seqan/core

File: include/seqan3/search/configuration/all.hpp [hunk 6]
```diff
@@ -128,4 +128,12 @@
...
  *
  * \include test/snippet/search/dynamic_hit_configuration_example.cpp
+ *
+ * \subsection search_configuration_subsection_parallel 6: Parallel Configuration
+ *
+ * This configuration determines the maximal number of threads the search algorithm can use.
+ *
+ * The seqan3::search_cfg::parallel configuration element can be combined with any other search configuration.
+ *
+ * \include test/snippet/search/configuration_parallel.cpp
```
Comment (smehringer) [inline include/seqan3/search/configuration/all.hpp:138]: Can you add an example to the snippet of how one can use the thread_count member?

File: include/seqan3/core/algorithm/configuration_element_parallel_mode.hpp (hunk 5 above)
Comment (joergi-w) [reply include/seqan3/core/algorithm/configuration_element_parallel_mode.hpp:39]: I added `noexcept` on purpose, because when I left it away, CLion told me:
```
Clang-Tidy: Move constructors should be marked noexcept.
Clang-Tidy: Move assignment operators should be marked noexcept.
```

If it's our policy I can remove them now.

File: include/seqan3/core/algorithm/configuration_element_parallel_mode.hpp (hunk 4 above)
Comment (joergi-w) [reply include/seqan3/core/algorithm/configuration_element_parallel_mode.hpp:45]: ok!

File: include/seqan3/core/algorithm/configuration_element_parallel_mode.hpp (hunk 5 above)
Comment (smehringer) [reply include/seqan3/core/algorithm/configuration_element_parallel_mode.hpp]: Yes :)
See here: https://en.cppreference.com/w/cpp/language/noexcept_spec that they are noexcept if they can.

**Comment by rrahn** [comment]
What should be the default thread count if the user does not specifies anything?

### Resolution 24.06.2020

The option should default to `std::nullopt` if default constructed and throws in the configuration if not set by the user explicitly.

**Comment by joergi-w** [comment]
I cannot assign `std::nullopt` to a `uint32_t` type. A solution is to make the thread_count variable of type `std::optional<uint32_t>` and then we have to access the variable as `config.thread_count.value` or `config->thread_count`. Is this intended or do you have another solution in mind?

**Comment by eseiler** [comment]
> I cannot assign `std::nullopt` to a `uint32_t` type. A solution is to make the thread_count variable of type `std::optional<uint32_t>` and then we have to access the variable as `config.thread_count.value` or `config->thread_count`. Is this intended or do you have another solution in mind?

Yes, `std::optional` was the intention

File: test/unit/alignment/pairwise/align_pairwise_test.cpp [hunk 7]
```diff
@@ -49,9 +50,9 @@ auto call_alignment(seq_t && seq, cfg_t && cfg)
     {
         return seqan3::align_pairwise(std::forward<seq_t>(seq), std::forward<cfg_t>(cfg));
     }
-    else
+    else if constexpr (std::same_as<type_param_t, seqan3::align_cfg::parallel>)
     {
-        auto && config = cfg | type_param_t{};
+        auto && config = cfg | seqan3::align_cfg::parallel{4};
```
Comment (smehringer) [inline test/unit/alignment/pairwise/align_pairwise_test.cpp:55]: I don't understand this change?

File: include/seqan3/alignment/pairwise/align_pairwise.hpp [hunk 8]
```diff
@@ -187,7 +188,7 @@ constexpr auto align_pairwise(sequence_t && sequences,
     auto select_execution_handler = [&] ()
     {
         if constexpr (std::same_as<execution_handler_t, detail::execution_handler_parallel>)
-            return execution_handler_t{get<align_cfg::parallel>(complete_config).value};
+            return execution_handler_t{get<align_cfg::parallel>(complete_config).thread_count.value()};
         else
             return execution_handler_t{};
```
Comment (smehringer) [inline include/seqan3/alignment/pairwise/align_pairwise.hpp:193]:
```suggestion
        {
            if (!get<align_cfg::parallel>(complete_config).thread_count.has_value())
                throw std::runtime_error{"You must configure the number of threads in seqan3::align_cfg::parallel."};
            return execution_handler_t{get<align_cfg::parallel>(complete_config).thread_count.value()};
        }
        else
        {
            return execution_handler_t{};
        }
```

I think this will be more easy to understand for the user than a `std::bad_optional_acces` ?
Maybe rephrase the sentence, it's just a proposal :)

File: test/unit/alignment/pairwise/align_pairwise_test.cpp (hunk 7 above)
Comment (joergi-w) [reply test/unit/alignment/pairwise/align_pairwise_test.cpp:55]: `type_param_t` is either void or `seqan3::align_cfg::parallel`, because we want to test the serial and the parallel version. And if it is parallel, we do not want to use the default initialisation, because half of the tests would fail with the std::bad_optional_access exception. So I initialise the parallel tests with 4.

File: include/seqan3/alignment/pairwise/align_pairwise.hpp (hunk 8 above)
Comment (joergi-w) [reply include/seqan3/alignment/pairwise/align_pairwise.hpp]: Yes, I like this suggestion :)

File: test/unit/alignment/pairwise/align_pairwise_test.cpp (hunk 7 above)
Comment (smehringer) [reply test/unit/alignment/pairwise/align_pairwise_test.cpp:55]: thank you for the explanation :)
